File: ios/Zest/ViewModels/CollaborativeMealPlanViewModel.swift
import Foundation
import Combine
import FirebaseAuth

class CollaborativeMealPlanViewModel: ObservableObject {
    @Published var collaborativeMealPlans: [CollaborativeMealPlan] = []
    @Published var selectedCollaborativeMealPlan: CollaborativeMealPlan?
    @Published var isLoading = false
    @Published var errorMessage: String?
    
    private let firebaseService = FirebaseService.shared
    private var cancellables = Set<AnyCancellable>()
    
    init() {
        loadCollaborativeMealPlans()
    }
    
    func loadCollaborativeMealPlans() {
        guard let userId = Auth.auth().currentUser?.uid else {
            errorMessage = "User not authenticated"
            return
        }
        isLoading = true
        errorMessage = nil
        Task { @MainActor in
            do {
                self.collaborativeMealPlans = try await firebaseService.getCollaborativeMealPlans(userId: userId)
            } catch {
                self.errorMessage = error.localizedDescription
            }
            self.isLoading = false
        }
    }
    
    func loadCollaborativeMealPlanDetail(id: String) {
        isLoading = true
        errorMessage = nil
        Task { @MainActor in
            do {
                self.selectedCollaborativeMealPlan = try await firebaseService.getCollaborativeMealPlan(id: id)
            } catch {
                self.errorMessage = error.localizedDescription
            }
            self.isLoading = false
        }
    }
    
    func createCollaborativeMealPlan(name: String) {
        guard let userId = Auth.auth().currentUser?.uid else {
            errorMessage = "User not authenticated"
            return
        }
        isLoading = true
        errorMessage = nil
        Task { @MainActor in
            do {
                let newPlan = CollaborativeMealPlan(name: name, ownerId: userId, members: [CollaborativeMealPlanMember(userId: userId, role: "owner")], mealPlanIds: [], createdAt: Date(), updatedAt: Date())
                let _ = try await firebaseService.createCollaborativeMealPlan(newPlan)
                self.loadCollaborativeMealPlans()
            } catch {
                self.errorMessage = error.localizedDescription
            }
            self.isLoading = false
        }
    }
    
    func addMemberToCollaborativeMealPlan(planId: String, member: CollaborativeMealPlanMember) {
        isLoading = true
        errorMessage = nil
        Task { @MainActor in
            do {
                try await firebaseService.addMemberToCollaborativeMealPlan(planId: planId, member: member)
                self.loadCollaborativeMealPlanDetail(id: planId)
            } catch {
                self.errorMessage = error.localizedDescription
            }
            self.isLoading = false
        }
    }
    
    func removeMemberFromCollaborativeMealPlan(planId: String, member: CollaborativeMealPlanMember) {
        isLoading = true
        errorMessage = nil
        Task { @MainActor in
            do {
                try await firebaseService.removeMemberFromCollaborativeMealPlan(planId: planId, member: member)
                self.loadCollaborativeMealPlanDetail(id: planId)
            } catch {
                self.errorMessage = error.localizedDescription
            }
            self.isLoading = false
        }
    }
    
    func addMealPlanToCollaborativePlan(planId: String, mealPlanId: String) {
        isLoading = true
        errorMessage = nil
        Task { @MainActor in
            do {
                try await firebaseService.addMealPlanToCollaborativePlan(planId: planId, mealPlanId: mealPlanId)
                self.loadCollaborativeMealPlanDetail(id: planId)
            } catch {
                self.errorMessage = error.localizedDescription
            }
            self.isLoading = false
        }
    }
    
    func removeMealPlanFromCollaborativePlan(planId: String, mealPlanId: String) {
        isLoading = true
        errorMessage = nil
        Task { @MainActor in
            do {
                try await firebaseService.removeMealPlanFromCollaborativePlan(planId: planId, mealPlanId: mealPlanId)
                self.loadCollaborativeMealPlanDetail(id: planId)
            } catch {
                self.errorMessage = error.localizedDescription
            }
            self.isLoading = false
        }
    }
}

File: android/app/src/main/java/com/neostudios/zest/ui/screens/CollaborativeMealPlanViewModel.kt
package com.neostudios.zest.ui.screens

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.google.firebase.auth.FirebaseAuth
import com.neostudios.zest.domain.model.CollaborativeMealPlan
import com.neostudios.zest.domain.model.CollaborativeMealPlanMember
import com.neostudios.zest.domain.repository.CollaborativeMealPlanRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

@HiltViewModel
class CollaborativeMealPlanViewModel @Inject constructor(
    private val collaborativeMealPlanRepository: CollaborativeMealPlanRepository,
    private val auth: FirebaseAuth
) : ViewModel() {

    private val _uiState = MutableStateFlow(CollaborativeMealPlanUiState())
    val uiState: StateFlow<CollaborativeMealPlanUiState> = _uiState

    init {
        loadCollaborativeMealPlans()
    }

    fun loadCollaborativeMealPlans() {
        _uiState.value = _uiState.value.copy(isLoading = true)
        viewModelScope.launch {
            auth.currentUser?.uid?.let { userId ->
                try {
                    val plans = collaborativeMealPlanRepository.getCollaborativeMealPlans(userId)
                    _uiState.value = _uiState.value.copy(
                        collaborativeMealPlans = plans,
                        isLoading = false
                    )
                } catch (e: Exception) {
                    _uiState.value = _uiState.value.copy(
                        error = e.localizedMessage,
                        isLoading = false
                    )
                }
            } ?: run {
                _uiState.value = _uiState.value.copy(
                    error = "User not authenticated",
                    isLoading = false
                )
            }
        }
    }

    fun loadCollaborativeMealPlanDetail(id: String) {
        _uiState.value = _uiState.value.copy(isLoading = true)
        viewModelScope.launch {
            try {
                val plan = collaborativeMealPlanRepository.getCollaborativeMealPlan(id)
                _uiState.value = _uiState.value.copy(
                    selectedCollaborativeMealPlan = plan,
                    isLoading = false
                )
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(
                    error = e.localizedMessage,
                    isLoading = false
                )
            }
        }
    }

    fun createCollaborativeMealPlan(name: String) {
        _uiState.value = _uiState.value.copy(isLoading = true)
        viewModelScope.launch {
            auth.currentUser?.uid?.let { userId ->
                try {
                    collaborativeMealPlanRepository.createCollaborativeMealPlan(name, userId)
                    loadCollaborativeMealPlans()
                } catch (e: Exception) {
                    _uiState.value = _uiState.value.copy(error = e.localizedMessage)
                } finally {
                    _uiState.value = _uiState.value.copy(isLoading = false)
                }
            }
        }
    }

    fun addMemberToCollaborativeMealPlan(planId: String, member: CollaborativeMealPlanMember) {
        viewModelScope.launch {
            try {
                collaborativeMealPlanRepository.addMemberToCollaborativeMealPlan(planId, member)
                loadCollaborativeMealPlanDetail(planId)
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(error = e.localizedMessage)
            }
        }
    }

    fun removeMemberFromCollaborativeMealPlan(planId: String, member: CollaborativeMealPlanMember) {
        viewModelScope.launch {
            try {
                collaborativeMealPlanRepository.removeMemberFromCollaborativeMealPlan(planId, member)
                loadCollaborativeMealPlanDetail(planId)
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(error = e.localizedMessage)
            }
        }
    }

    fun addMealPlanToCollaborativePlan(planId: String, mealPlanId: String) {
        viewModelScope.launch {
            try {
                collaborativeMealPlanRepository.addMealPlanToCollaborativePlan(planId, mealPlanId)
                loadCollaborativeMealPlanDetail(planId)
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(error = e.localizedMessage)
            }
        }
    }

    fun removeMealPlanFromCollaborativePlan(planId: String, mealPlanId: String) {
        viewModelScope.launch {
            try {
                collaborativeMealPlanRepository.removeMealPlanFromCollaborativePlan(planId, mealPlanId)
                loadCollaborativeMealPlanDetail(planId)
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(error = e.localizedMessage)
            }
        }
    }
}

data class CollaborativeMealPlanUiState(
    val collaborativeMealPlans: List<CollaborativeMealPlan> = emptyList(),
    val selectedCollaborativeMealPlan: CollaborativeMealPlan? = null,
    val isLoading: Boolean = false,
    val error: String? = null
)
